jzaia_files/main.cpp

```diff
@@ -0,0 +1,225 @@
+ *
+ * @param sv The statevector to apply the oracle to. Must be 6 qubits
+ * @param control_vals The state the oracle should select
+ */
+void oracle(StateVectorLQubitManaged<double> &sv,
+            const std::vector<bool> &control_vals) {
+    // Sanity check statevector
+    assert(sv.getNumQubits() == control_vals.size()+1);
+
+    // Define controls to be used for applying the X gate
+    std::vector<size_t> controls(control_vals.size());
+    std::iota(controls.begin(), controls.end(), 0);
+
+    // Apply the X gate to the ancilla, controlled on the chosen bitstring
+    GateImplementationsLM::applyNCPauliX(sv.getData(),
+                                         sv.getNumQubits(),
+                                         controls,
+                                         control_vals,
+                                         {control_vals.size()},
+                                         false);
+}
+
+/**
+ * @brief Overall function for running Grover's algorithm on a chosen oracle
+ *
+ * Run Grover's algorithm from start to finish. Prepares a statevector, and
+ * repeats state selection and amplitude-amplification for sqrt(N) iterations
+ * (where N = 2^(# of non-ancilla qubits)). This implementation assumes that
+ * the oracle always picks precisly 1 state (rather than an arbitrary number).
+ *
+ * @param num_qubits The number of qubits in the circuit the oracle acts
+ *        on (includes the ancilla)
+ * @param expected The state which the oracle should apply phase to
+ */
+void run_experiment(const size_t num_qubits,
+                    const std::vector<bool> &expected) {
+    // Create the statevector for this circuit
+    // NOTE: qubits are numbered left-to-right order
+    StateVectorLQubitManaged<double> sv(num_qubits);
+
+    // Set up the circuit by placing into uniform superposition
+    groversSetup(sv);
+
+    // Apply (approx) sqrt(N) repetitions of state selection and amp-amp
+    const size_t nreps = static_cast<size_t>(sqrt(1llu << (num_qubits-1)));
+    for (size_t reps = nreps; reps > 0; --reps) {
+        // Apply the oracle to apply a phase of -1 to desired state
+        oracle(sv, expected);
+        // Perform amp-amp by reflecting over |+++...+>
+        groversMirror(sv);
+    }
+
+    // Set up measurements
+    Measurements<StateVectorLQubitManaged<double>> Measurer(sv);
+    // Vector to store the most common measurement outcome
+    std::vector<size_t> wires(num_qubits - 1);
+    std::iota(wires.begin(), wires.end(), 0);
+    std::vector<bool> common_result(num_qubits - 1, false);
+
+    // Perform a "measurement" by taking the expected value of this qubit over multiple runs
+    std::transform(wires.begin(),
+                   wires.end(),
+                   common_result.begin(),
+                   [&Measurer](size_t wire){
+                       NamedObs<StateVectorLQubitManaged<double>> obs("PauliZ", {wire});
+                       return Measurer.expval(obs) < 0;
+                   });
```

Comment:
Could you execute this same logic without creating and populating an extra vector `wires`?